From d074ca27259a106f435fce037315726a8de3e9f2 Mon Sep 17 00:00:00 2001
Message-Id: <d074ca27259a106f435fce037315726a8de3e9f2.1417780177.git.chang-joon.lee@intel.com>
In-Reply-To: <e293659a31a2970805251d9fed030ac809a9a6d0.1417780177.git.chang-joon.lee@intel.com>
References: <e293659a31a2970805251d9fed030ac809a9a6d0.1417780177.git.chang-joon.lee@intel.com>
From: Adam Rutkowski <adam.j.rutkowski@intel.com>
Date: Mon, 25 Aug 2014 07:24:53 -0400
Subject: [PATCH 26/30] REVERTME [VPG]: drm/i915: Perfmon open/close

Implementing Perfmon open/close IOCTL. Sysadmin capability
is required to open Perfmon in context of current fd.
Perfmon must be opened in order to set global perfmon
configuration and to get list of HW context IDs for arbitrary
context.

Patch is marked with REVERTME as Perfmon IOCTL will be removed
once OA couters are exposed via 'perf event' framework.

Issue: GMINL-3120
Change-Id: I20f1275501311a876d571329cea8039de3a0f384
Signed-off-by: Adam Rutkowski <adam.j.rutkowski@intel.com>
Signed-off-by: Andrzej Datczuk <andrzej.datczuk@intel.com>

Signed-off-by: CJ Lee <chang-joon.lee@intel.com>
---
 drivers/gpu/drm/i915/i915_drv.h          |    2 ++
 drivers/gpu/drm/i915/i915_perfmon.c      |   38 ++++++++++++++++++++++++++++--
 drivers/gpu/drm/i915/i915_perfmon_defs.h |    4 ++++
 3 files changed, 42 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 0dffa71..9e5ee78 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -2096,6 +2096,8 @@ struct drm_i915_file_private {
 	atomic_t rps_wait_boost;
 	struct  intel_engine_cs *bsd_ring;
 	struct bin_attribute *obj_attr;
+
+	struct drm_i915_perfmon_file perfmon;
 };
 
 /*
diff --git a/drivers/gpu/drm/i915/i915_perfmon.c b/drivers/gpu/drm/i915/i915_perfmon.c
index 5d2cba2..007085e 100644
--- a/drivers/gpu/drm/i915/i915_perfmon.c
+++ b/drivers/gpu/drm/i915/i915_perfmon.c
@@ -114,6 +114,40 @@ static int intel_cancel_wait_perfmon_interrupt(struct drm_device *dev)
 }
 
 /**
+ * i915_perfmon_open
+ *
+ * open perfmon for current file
+ */
+static int i915_perfmon_open(
+	struct drm_file *file)
+{
+	struct drm_i915_file_private *file_priv = file->driver_priv;
+	int ret = 0;
+
+	if (!capable(CAP_SYS_ADMIN))
+		ret = -EACCES;
+	else
+		file_priv->perfmon.opened = true;
+
+	return ret;
+}
+
+/**
+ * i915_perfmon_close
+ *
+ * close perfmon for current file
+ */
+static int i915_perfmon_close(
+	struct drm_file *file)
+{
+	struct drm_i915_file_private *file_priv = file->driver_priv;
+
+	file_priv->perfmon.opened = false;
+
+	return 0;
+}
+
+/**
  * i915_perfmon_ioctl - performance monitoring support
  *
  * Main entry point to performance monitoring support
@@ -148,11 +182,11 @@ int i915_perfmon_ioctl(struct drm_device *dev, void *data,
 		break;
 
 	case I915_PERFMON_OPEN:
-		ret = -ENODEV;
+		ret = i915_perfmon_open(file);
 		break;
 
 	case I915_PERFMON_CLOSE:
-		ret = -ENODEV;
+		ret = i915_perfmon_close(file);
 		break;
 
 	case I915_PERFMON_ENABLE_CONFIG:
diff --git a/drivers/gpu/drm/i915/i915_perfmon_defs.h b/drivers/gpu/drm/i915/i915_perfmon_defs.h
index 6509aad..169d2e8 100644
--- a/drivers/gpu/drm/i915/i915_perfmon_defs.h
+++ b/drivers/gpu/drm/i915/i915_perfmon_defs.h
@@ -30,4 +30,8 @@ struct drm_i915_perfmon_device {
 	atomic_t		buffer_interrupts;
 };
 
+struct drm_i915_perfmon_file {
+	bool opened;
+};
+
 #endif	/* _I915_PERFMON_DEFS_H_ */
-- 
1.7.9.5

